splitgpg2/__init__.py

```diff
@@ -0,0 +1,804 @@
+        self.agent_socket_path = agent_socket_path.decode()
+
+        loop = asyncio.get_event_loop()
+        self.agent_transport, self.agent_protocol = await \
+            loop.create_unix_connection(lambda: AgentProtocol(self),
+                                        self.agent_socket_path)
+        # wait for agent hello
+        await self.agent_protocol.agent_responded
+        self.state = ServerState.client_command
+
+    def connection_made(self, transport: asyncio.BaseTransport) -> None:
+        self.client_transport = transport
+        asyncio.ensure_future(self.connect_agent())
+
+    def connection_lost(self, exc: Optional[Exception]) -> None:
+        for fut in self.notify_on_disconnect:
+            fut.set_result(None)
+
+    def abort(self, reason):
+        self.log.error('%s; Aborting!', reason)
+        self.client_transport.close()
+        self.agent_transport.close()
+
+    async def handle_command(self, *, untrusted_line: bytes):
+        # state == ServerState.client_command
+        try:
+            try:
+                untrusted_cmd, untrusted_args = untrusted_line.split(b' ', 1)
+            except ValueError:
+                untrusted_cmd = untrusted_line
+                untrusted_args = None
+            try:
+                command = self.commands[untrusted_cmd]
+            except KeyError:
+                raise Filtered
+            await command(untrusted_args=untrusted_args)
+        except Filtered as e:
+            # DEBUG
+            self.log.exception('Filtered')
+
+            self.notify('command filtered out')
+            self.client_write('ERR {} {}\n'.format(e.code, e.gpg_message).encode())
+            # break handling since we aren't sure that clients handle the error
+            # correctly. This makes the filtering easier to implement and we ensure
+            # that a client does not wrongly assume that a command was successful
+            # while is was indeed filtered out.
+            self.abort('command filtered out')
+        except:  # pylint: disable=bare-except
+            self.log.exception('Error processing command')
+            self.abort('error')
+
+    async def handle_inquire(self, *, untrusted_line: bytes):
+        # state == ServerState.client_inquire
+        try:
+            try:
+                untrusted_cmd, untrusted_args = untrusted_line.split(b' ', 1)
+            except ValueError:
+                untrusted_cmd = untrusted_line
+                untrusted_args = None
+            try:
+                inquire_command = self.inquire_commands[untrusted_cmd]
+            except KeyError:
+                raise Filtered
+            cont = await inquire_command(untrusted_args=untrusted_args)
+            if not cont:
+                self.inquire_complete.set_result(None)
+        except Filtered as e:
+            self.notify('command filtered out')
+            self.client_write('ERR {} {}\n'.format(e.code, e.gpg_message).encode())
```

Comment:
This is duplicated, better move it into a fucntion.

Reply:
ok